scanner/grpc/server.go

```diff
@@ -0,0 +1,155 @@
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"time"
+
+	pb "github.com/Layr-Labs/eigenda/api/grpc/scanner"
+	"github.com/Layr-Labs/eigenda/common"
+	"github.com/Layr-Labs/eigenda/scanner"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"github.com/Ullaakut/nmap"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/reflection"
+)
+
+const localhost = "0.0.0.0"
+
+// Server implements the Scanner proto APIs.
+type Server struct {
+	pb.UnimplementedReachabilityServer
+
+	scanner *scanner.Scanner
+	config  *scanner.Config
+	logger  logging.Logger
+
+	ratelimiter common.RateLimiter
+}
+
+// NewServer func
+func NewServer(config *scanner.Config, scanner *scanner.Scanner, logger logging.Logger) *Server {
+	return &Server{
+		config:      config,
+		logger:      logger,
+		scanner:     scanner,
+		ratelimiter: nil,
+	}
+}
+
+// Start func
+func (s *Server) Start() error {
+
+	addr := fmt.Sprintf("%s:%s", localhost, s.config.ServerPort)
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Fatalf("failed to start tcp listener: %v", err)
+	}
+
+	s.logger.Info("registering service")
+	//opt := grpc.MaxRecvMsgSize(128)
+	gs := grpc.NewServer()
+	reflection.Register(gs)
+	pb.RegisterReachabilityServer(gs, s)
+
+	s.logger.Info("port", s.config.ServerPort, "address", lis.Addr().String(), "GRPC Listening")
+	if err := gs.Serve(lis); err != nil {
+		log.Fatalf("failed to serve: %v", err)
+	}
+	return nil
+}
+
+// PortCheck func
+func (s *Server) PortCheck(ctx context.Context, in *pb.PortCheckRequest) (*pb.PortCheckResponse, error) {
+
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		log.Println("Could not get peer from context")
+		return &pb.PortCheckResponse{Status: "error", Msg: "Could not get peer ip from context"}, nil
+	}
+	peerAddr := p.Addr.String()
+	ip, _, err := net.SplitHostPort(peerAddr)
+	if err != nil {
+		log.Printf("Could not split host and port: %v", err)
+		return &pb.PortCheckResponse{Status: "error", Msg: "Could not parse host ip"}, nil
+	}
+
+	var timeout = 5 * time.Second
+	_, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	scan, err := nmap.NewScanner(
```

Comment:
Not sure how `nmap` works, but does this succeed if the port accepts grpc requests but not the other kinds of requests?

Reply:
Nmap is simply checking if the port accepts connection at OSI layer 3-4 (ie TCP/IP)